src/main/scala/scala/slick/jdbc/TypedStaticQuery.scala

```diff
@@ -0,0 +1,378 @@
+          createClassTreeFromString("scala.slick.jdbc.TypedStaticQuery.ConfigHandler", newTypeName(_))
+        ), nme.CONSTRUCTOR),
+      List(Literal(Constant(config.dbName)))
+    )
+    
+    /**
+     * Create a Tree of the static name of a class
+     * eg java.lang.String becomes
+     * Select(Select(Select(Ident(nme.ROOTPKG), "java"), "lang"), "String")
+     */
+    def createClassTreeFromString(classString: String, generator: String => Name): Tree = {
+      val tokens = classString.split('.').toList
+      val packages = tokens.dropRight(1) map (newTermName(_))
+      val classType = generator(tokens.last)
+      val firstPackage = Ident(nme.ROOTPKG)
+      val others = (packages :+ classType)
+      others.foldLeft[Tree](firstPackage)((prev, elem) => {
+        Select(prev, elem)
+      })
+    }
+    
+    //Shorthand for c.eval
+    private[this] def eval[T](tree: Tree): T = c.eval(c.Expr[T](c.resetLocalAttrs(tree)))
+
+    /**
+     * Actually locates a TSQLConfig annotation and
+     * creates a Tree of ConfigHandler
+     */
+    lazy val configHandlerTree: Tree = {
+
+      //From a list of annotations determine the TSQLConfig annotation
+      def findAnnotationTree(ann: List[Tree]): Option[Tree] = ann.flatMap { tree =>
+        c.typeCheck(tree, pt = weakTypeOf[TSQLConfig], silent = true) match {
+          case EmptyTree => None
+          case _ => {
+            val Apply(Select(_, _), args) = tree
+            val realTree = Apply(
+              Select(
+                New(createClassTreeFromString("scala.slick.jdbc.TypedStaticQuery.TSQLConfig", newTypeName(_))),
+                nme.CONSTRUCTOR
+              ), args map (_.duplicate)
+            )
+            Some(realTree)
+          }
+        }
+      }.headOption
+
+      //Determine the trees
+      val clasDef = c.enclosingClass.asInstanceOf[MemberDef]
+      val methDef = Option(c.enclosingMethod).filter(_ != EmptyTree).map(_.asInstanceOf[MemberDef])
+
+      //Determine the annotations and evaluate corresponding ConfigHandlers
+      val clasConf = findAnnotationTree(clasDef.mods.annotations) map {t =>
+        createConfigHandler(eval[TSQLConfig](t))
+      }
+      val methConf = methDef.flatMap(md => findAnnotationTree(md.mods.annotations) map {t =>
+        createConfigHandler(eval[TSQLConfig](t))
+      })
+
+      methConf.getOrElse {
+        clasConf.getOrElse{
+          abort("Cannot find suitable config handler for this invocation")
+        }
+      }
+    }
+
+    /**
+     * Actually locates a TSQLConfig annotation and
+     * creates an instance of ConfigHandler
+     */
+    lazy val configHandler: ConfigHandler = eval[ConfigHandler](configHandlerTree)
+  }
+
+  /**
+   * The class that is used to ensure an uniform database access mechanism
+   * at the compile-time and run-time by factoring the connection parameters
+   */
+  final class ConfigHandler(databaseName: String) {
+    
+    final val configFileName = "reference.conf"
```

Comment:
I have a branch with a higher-level configuration abstraction that can be used for the code generator, the tsql macro, and probably also play-slick. We decided to leave the config stuff in this PR untouched because I'm still working on integrating it properly.